fix(stepfunctions): incorrect/missing permissions to run/redrive DistributedMap in state machine

Kasra-G opened the pull request:
### Issue # (if applicable)

Closes #35390.

### Reason for this change

<!--What is the bug or use case behind this change?-->
When using the `DistributedMap` state, the state machine cannot be redriven without additional configuration from the user. These additional steps seem to sometimes lead users into circular dependency issues (see [example](https://github.com/aws/aws-cdk/issues/28820#issuecomment-2065316882)). 

Additionally, the `DistributedMap` assigns incorrect permissions for the `states:StopExecution` and `states:DescribeExecution` as these need to be applied to the execution arn, not the state machine arn ([AWS Docs](https://docs.aws.amazon.com/step-functions/latest/dg/iam-policies-eg-dist-map.html) vs [code reference](https://github.com/aws/aws-cdk/blob/aea1372ab7bc68c489cea5ee5e499233755910e8/packages/aws-cdk-lib/aws-stepfunctions/lib/state-graph.ts#L178-L180).

This PR also fixes an inconsistency for the recently added `grantRedriveExecution` function as it does not work for state machines that have DistributedMaps. 
The root cause is that when adding a DistributedMap to a state machine, additional permissions need to be added to the state machine's execution role to permit redriving a map run. Otherwise, when the state machine is redriven, the map run will fail to redrive. The docs provide a minimum permissions example https://docs.aws.amazon.com/step-functions/latest/dg/iam-policies-eg-dist-map.html#iam-policy-redrive-dist-map
The key permission is `states:RedriveExecution` on the following arn:
`arn:aws:states:us-east-2:123456789012:execution:myStateMachineName/myMapRunLabel:*`

However, when creating a state machine manually in the AWS Console, if there are any unlabeled distributed maps, the following policy is generated for the generated state machine role:
`arn:aws:states:us-east-2:123456789012:execution:myStateMachineName/*:*`

This change mirrors that logic in CDK.

### Description of changes

Minor: fix `map-with-catch` integ test failing because it was not emptying the bucket before attempting to delete it

- Updated the permissions granted in the `bind` method for DistributedMaps to include correct redrive permissions
- Fixed incorrect resource on the `states:StopExecution` and `states:DescribeExecution` permissions
- ~~Fix missing DistributedMap policy statements when the DistributedMap state was in a child `StateGraph` such as a `Parallel.branch`~~ This will be handled in a separate PR.

These changes fix the root cause of `grantRedriveExecution` not working for state machines with DistributedMaps

### Describe any new or updated permissions being added

when there is any unlabeled distributed map run state
```diff
  action: ["states:DescribeExecution", "states:StopExecution"]
- resource: "arn:aws:states:us-east-2:account-id:stateMachine:myStateMachine:*"
+ resource: "arn:aws:states:us-east-2:account-id:execution:myStateMachine:*"
//...
+ action: "states:RedriveExecution"
+ resource: "arn:aws:states:us-east-2:account-id:execution:myStateMachine/*:*"
```
when there are only labeled distributed map runs
```diff
  action: ["states:DescribeExecution", "states:StopExecution"]
- resource: "arn:aws:states:us-east-2:account-id:stateMachine:myStateMachine:*"
+ resource: "arn:aws:states:us-east-2:account-id:execution:myStateMachine:*"
//...
+ action: "states:RedriveExecution"
+ resource: [
+   "arn:aws:states:us-east-2:account-id:execution:myStateMachine/myLabel1:*", 
+   "arn:aws:states:us-east-2:account-id:execution:myStateMachine/myLabel2:*"
+ ]
```

### Description of how you validated changes

Unit & Integration tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (dismissed):
(This review is outdated)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3720ec0dcd46b19d01cf2cd683449017b1c6a071
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=790c0d6d-5d06-48ef-a748-7edb46c1f24f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d65f21dbc2132673e8951677bc429e14d47633fc
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e097c937-90c7-4392-8fa5-d93daa82d80b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Kasra-G commented:
I think I would rather rewrite this to not change the `grantRedriveExecution` permissions as that technically doesn't need to change as long as the state machine has the permissions to redrive execution on all of its own labeled executions

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a03f4d422fb2d732207bce85449a151aedb7257a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c5d8192f-3a9e-4896-9cee-2de5f004c45e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8b6afd4265bd16aba40fc5adf1d547e01abec408
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a06c7560-a3f2-478e-9b47-b78f26ea1c59%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kumvprat commented on packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions/test/integ.distributed-map-redrive.ts:
```diff
@@ -0,0 +1,118 @@
+
+// assert the results
+describe
+  .expect(ExpectedResult.objectLike({ status: 'FAILED' }))
```
I am a bit confused here : How do we ensure that the state function fails on first try and execute a redrive of the state function? Would be good to call it out explicitly

kumvprat reviewed (commented):
@Kasra-G Thanks for the contribution, this seems to be a niche one
I will add inline comments for this

Since it has been a while since the issue was reported, I am not sure if having the RedriveExecution policy on the DistributedMap state type is a default expectation(maybe it's called out in the documentation and I missed it, if so would be nice to link it in the PR description)

There are integration tests which cover the basic distributed map functionality : `packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions/test/integ.distributed-map.ts` which were working before(?)without explictly adding the `RedriveExecution` based policy. (trying to make a case for not needing the `RedriveExecution` policy by default)

Kasra-G commented:
Thanks for taking a look at the PR. I have updated it and have a few responses:

> Since it has been a while since the issue was reported, I am not sure if having the RedriveExecution policy on the DistributedMap state type is a default expectation(maybe it's called out in the documentation and I missed it, if so would be nice to link it in the PR description)

I have updated the PR description with link to AWS docs talking about the permissions needed when adding a DistributedMap to a state machine. 

> There are integration tests which cover the basic distributed map functionality : `packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions/test/integ.distributed-map.ts` which were working before(?)without explictly adding the `RedriveExecution` based policy. 

Yes, those integ tests work, but they do not test the redrive capability. That is why I have added a new integration test that tests the redrive capability for a state machine with a distributed map. 

> (trying to make a case for not needing the `RedriveExecution` policy by default)

The redrive permissions are added when creating a distributed map in a state machine via the AWS Console (and selecting the option to create a new role), but not in CDK

Kasra-G commented:
Sidenote: any idea why PR Build workflow isn't running?

Kasra-G commented:
I also have changes that would fix the same underlying issue as https://github.com/aws/aws-cdk/pull/29913 before it was closed for inactivity. I was going to open another PR for that one after this one, but since the changes conflict, I can merge the changes into the same PR to reduce overhead - just let me know.

Kasra-G commented on packages/aws-cdk-lib/aws-stepfunctions/lib/state-graph.ts:
```diff
@@ -163,28 +163,69 @@ export class StateGraph {
+  /**
+   * Returns all the states referenced in this StateGraph and all child StateGraphs
+   */
+  private getAllStatesInAllReferencedStateGraphs(): State[] {
```
Sure, I can separate it out, though comments on the reference issue cite distributed maps inside parallel states not running because of missing permissions https://github.com/aws/aws-cdk/issues/28820#issuecomment-3050416978. I figured that could be in scope of this PR given it aims to fix these policies

I will then create a new issue and reference that in this PR to prevent the main discussion issue from being closed while the other PR is reviewed.

Kasra-G commented on packages/aws-cdk-lib/aws-stepfunctions/lib/state-graph.ts:
```diff
@@ -163,28 +163,69 @@ export class StateGraph {
+    const discoveredStateGraphs = new Set<StateGraph>();
+    const stateGraphsToSearch: StateGraph[] = [this];
+
+    // Do a DFS of all sub graphs
```
True, I will update. I had BFS at first but forgot which side of the array `push()` puts the elements in. Either way theres no real difference as we are just trying to find all reachable states.

kumvprat commented on packages/aws-cdk-lib/aws-stepfunctions/lib/state-graph.ts:
```diff
@@ -163,28 +163,69 @@ export class StateGraph {
+    stateMachine.role.attachInlinePolicy(distributedMapPolicy);
+    stateMachine.grantStartExecution(distributedMapPolicy);
+    stateMachine.grantExecution(distributedMapPolicy, 'states:DescribeExecution', 'states:StopExecution');
+
```
Thanks for sharing this. I think the main thing we want is parity in how the sfn permissions are created in AWS console and in cdk

kumvprat commented on packages/aws-cdk-lib/aws-stepfunctions/test/state-machine.test.ts:
```diff
@@ -173,16 +187,375 @@ describe('State Machine', () => {
+    });
+  }),
+
+  test('Instantiate State Machine With Self Referencing Distributed Map State', () => {
```
That makes sense, we do want to point out why we are testing it. Customers would be free to use the functionality if it's available.

A comment on the test would help us understand in future why is this being tested out

Kasra-G commented:
> Are the changes to just the index.js for aws-lambda-python-alphs and aws-ses packages integ tests needed

Probably not. I think those integ tests failed on my first version of the PR so I ran them to update the files but there wasn't really any changes. I'll try resetting those changes.

Kasra-G commented:
Apologies, just noticed I have committed snapshot files for an integ test that I renamed all the way back in my first or second pr revision. I will remove those snapshot files.

kumvprat reviewed (approved):
Thank you for detailed work on the PR
